CDAP-14442 refactor file sink formats into plugins

albertshau opened the pull request:
This will fix the classloading issues we have when plugins from
different artifacts are using avro or parquet.

albertshau committed:
CDAP-14442 refactor file sink formats into plugins

This will fix the classloading issues we have when plugins from
different artifacts are using avro or parquet.

albertshau commented:
https://builds.cask.co/browse/HYP-BAD499

chtyim commented on format-avro/src/main/java/co/cask/format/avro/AvroOutputFormatProvider.java:
```diff
@@ -0,0 +1,107 @@
+    if (conf.compressionCodec != null && !conf.containsMacro("compressionCodec") &&
+      !"none".equalsIgnoreCase(conf.compressionCodec)) {
+
+      configuration.put(MAPRED_OUTPUT_COMPRESS, "true");
```
Better use the `CodecFactory` class from the avro library to parse and pickup the codec to simplify the logic.

```java
try {
  CodecFactory.fromString(conf.compressionCodec);
} catch (Exception e) {
  throw new IllegalArgumentException(.....);
}

chtyim commented on format-parquet/pom.xml:
```diff
@@ -0,0 +1,105 @@
+            <_exportcontents>
+              co.cask.format.parquet.*;
+              org.apache.parquet.avro.*;
+              org.apache.parquet.hadoop.*;
```
Why need to expose the parquet? I thought we wrap the input/output format class?

albertshau replied:
oh true, I had just copied these over without thinking. I'll try removing these.

albertshau replied:
There are some classloading issues if I remove these, I think because of the fact that they haven't been removed from core-plugins yet since I haven't moved the inputformats out yet.

chtyim replied:
I see. So I suppose these will be removed when the inputformat/source modification is done.

albertshau committed:
use avro and parquet code for determining codec validity

Also fix a bug with tpfs and snapshot sinks where datasets created
by a pervious version of the plugin would not be readable, due
to differences in output format stored in the dataset properties.
Changed the sinks to override any properties from dataset creation
with those that the plugin knows are needed at runtime.
